### DISCUSSION TITLE
user functions - best location and best practice for import?

### DISCUSSION THREAD

**Discussion by Munchotaur** [xonsh/xonsh#5045] [General]
What's the best practice for auto-loading user written functions at xonsh startup?

Ideally I'd like to imitate fish's .config/<name_of_shell>/functions directory and auto-load whatever .py modules I put in there, but is there a better/more standard way?

**Comment by gforsyth** [comment]
You could put the following in your `xonshrc`:

```
for f in g`~/.config/xonsh/functions/*`:
    source @(f)
```

and that would source every `xsh` and `py` file in `functions`

**Comment by anki-code** [comment]
[Run Control File](https://xon.sh/xonshrc.html):

> Xonsh also supports configuration directories, from which all .xsh files will be sourced in order. This allows for drop-in configuration where your configuration can be split across scripts and common and local configuration more easily separated. By default, if the directory **~/.config/xonsh/rc.d** exists, any *.xsh files within will be sourced at startup.

**Comment by Munchotaur** [comment]
Thanks gforsyth, anki-code. I'll give those a try.

**Reply by anki-code** [reply]
I also recommend to read [xonsh-cheatsheet](https://github.com/anki-code/xonsh-cheatsheet/).

**Reply by Munchotaur** [reply]
Will do. Thanks.
